megazords/ios/RustLog/RustLog.swift

```diff
@@ -0,0 +1,247 @@
+            self = result
+            return
+        }
+
+        if value < LogLevel.trace.rawValue {
+            self = .trace
+        } else {
+            self = .error
+        }
+    }
+}
+
+/// An enum representing a maximum log level. It is used with
+/// `RustLog.shared.setLevelFilter`.
+///
+/// This is roughly equivalent to LogLevel, however contains
+/// `Off`, for filtering all logging.
+public enum LogLevelFilter: Int32 {
+    /// Disable all logging
+    case off
+    /// Only allow `error` logs.
+    case error
+    /// Allow `warn` and `error` logs.
+    case warn
+    /// Allow `warn`, `error`, and `info` logs.
+    case info
+    /// Allow `warn`, `error`, `info`, and `debug` logs. The default.
+    case debug
+    /// Allow all logs, including those that may contain PII.
+    case trace
+}
+
+/// The type of the log callback. You can provide a value of this type to
+/// `RustLog.shared.enable` or `RustLog.shared.tryEnable`, and it will be called for
+/// all log messages emitted by Rust code.
+///
+/// The first argument is the level of the log. The maximum value of this can
+/// be provided using the `RustLog.shared.setLevelFilter` method.
+///
+/// The second argument is the tag, which is typically a rust module path
+/// string. It might be nil in some cases that aren't documented by the
+/// underlying rust log crate.
+///
+/// The last argument is the log message. It will not be nil.
+///
+/// This callback should return `true` to indicate everything is fine, and
+/// false if we should disable the logger. You cannot call `disable()`
+/// from inside the callback (it's protected by a dispatch queue you're
+/// already running on).
+public typealias LogCallback = (_ level: LogLevel, _ tag: String?, _ message: String) -> Bool;
+
+/// The public interface to Rust's logger.
+///
+/// This is a singleton, and should be used via the
+/// `shared` static member.
+public class RustLog {
+    fileprivate let state = RustLogState()
+    fileprivate let queue = DispatchQueue(label: "com.mozilla.appservices.rust-log")
+    /// The singleton instance of RustLog
+    public static let shared = RustLog()
+
+    private init() {}
+
+    /// True if the logger currently has a bound callback.
+    public var isEnabled: Bool {
+        return queue.sync { state.isEnabled }
+    }
+
+    /// Set the current log callback.
+    ///
+    /// Note that by default, after enabling the level filter
+    /// will be at the `debug` level. If you want to increase or decrease it,
+    /// you may use `setLevelFilter`
+    ///
+    ///
+    /// See alse `tryEnable`.
+    ///
+    /// Throws:
+    ///
+    /// - `RustLogError.alreadyEnabled`: If we're already enabled. Explicitly disable first.
+    ///
+    /// - `RustLogError.unexpectedError`: If the rust code panics. This shouldn't happen,
+    ///   but if it does, we would appreciate reports from telemetry or similar
+    public func enable(_ callback: @escaping LogCallback) throws {
```

Comment:
Oh, I see now. `enable` takes a callback, so, once it's disabled, we can re-enable by passing a new callback.